File: VaultTrackBackend/src/main/java/com/example/VaultTrackBackend/model/entity/RecurringTransaction.java
package com.example.VaultTrackBackend.model.entity;

import com.example.VaultTrackBackend.model.enums.RecurringFrequency;
import com.example.VaultTrackBackend.model.enums.TransactionCategory;
import com.example.VaultTrackBackend.model.enums.TransactionType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.GenericGenerator;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "recurring_transactions")
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class RecurringTransaction {

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "recurring_transaction_id", updatable = false, nullable = false)
    private UUID recurringTransactionId;

    @Column(name = "transaction_name", nullable = false)
    private String transactionName;

    @Column(name = "amount", nullable = false)
    private BigDecimal amount;

    @Column(name = "transaction_category", nullable = false)
    @Enumerated(EnumType.STRING)
    private TransactionCategory transactionCategory;

    @Column(name = "transaction_type", nullable = false)
    @Enumerated(EnumType.STRING)
    private TransactionType transactionType;

    @Column(name = "recurrence_frequency", nullable = false)
    @Enumerated(EnumType.STRING)
    private RecurringFrequency recurringFrequency;

    @Column(name = "next_execution_date", nullable = false)
    private LocalDate nextExecutionDate;

    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne
    @JoinColumn(name = "account_id", nullable = false)
    private Account account;
}

File: VaultTrackFrontend/src/types/recurringTransaction.ts
import { TransactionCategory, TransactionType } from './transaction';

export enum RecurringFrequency {
  DAILY = 'DAILY',
  WEEKLY = 'WEEKLY',
  MONTHLY = 'MONTHLY',
  YEARLY = 'YEARLY'
}

export interface RecurringTransaction {
  recurringTransactionId: string;
  transactionName: string;
  amount: number;
  transactionCategory: TransactionCategory;
  transactionType: TransactionType;
  recurringFrequency: RecurringFrequency;
  nextExecutionDate: string; // YYYY-MM-DD format
  isActive: boolean;
  accountId: string;
  accountName: string;
  createdAt: string;
}

export interface CreateRecurringTransactionData {
  transactionName: string;
  amount: number;
  transactionCategory: TransactionCategory;
  transactionType: TransactionType;
  recurringFrequency: RecurringFrequency;
  nextExecutionDate: string; // YYYY-MM-DD format
  accountId: string;
}

export interface UpdateRecurringTransactionData {
  transactionName: string;
  amount: number;
  transactionCategory: TransactionCategory;
  transactionType: TransactionType;
  recurringFrequency: RecurringFrequency;
  nextExecutionDate: string; // YYYY-MM-DD format
  accountId: string;
}
